### PR TITLE
TDL-15922 Implement Request Timeouts

### PR DESCRIPTION
**Pull request by prijendev**
# Description of change
- Implemented request timeouts.
- Set timeout to config parameter `request_timeout` value.
- If config parameter does not pass or pass the value "", "0", 0, "0.0" then set timeout to default value 5min

# Manual QA steps
 - Test that tap perform backoff on timeout error.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/unittests/test_support_eu_endpoints.py [hunk 1]
```diff
@@ -77,13 +77,15 @@ def test_support_eu_endpoints_except_export(self, mock_write_schema, mock_write_
         revenue_obj = Revenue(client)
         revenue_obj.sync(catalog=catalog, state=state, config=EU_CONFIG, start_date="2020-02-01T00:00:00Z")
 
-        mock_request.assert_called_with(method='GET', url='https://eu.mixpanel.com/api/2.0', path='engage/revenue',
+        # Verify that sync method called with expected parameter
+        mock_request.assert_called_with(method='GET', request_timeout=300, url='https://eu.mixpanel.com/api/2.0', path='engage/revenue',
```
Comment (dbshah1212) [inline tests/unittests/test_support_eu_endpoints.py:81]: Please use constant here for 300

File: .circleci/config.yml [hunk 2]
```diff
@@ -43,6 +44,10 @@ jobs:
             source /usr/local/share/virtualenvs/tap-tester/bin/activate
             PYTHONPATH=$PYTHONPATH:/usr/local/share/virtualenvs/tap-mixpanel/lib/python3.5/site-packages/ \
             run-test --tap=tap-mixpanel tests/tap_tester
+      - store_test_results:
```
[the remark this answers is not in the document]
Comment (prijendev) [reply .circleci/config.yml]: updated

File: README.md [hunk 3]
```diff
@@ -138,6 +138,7 @@ More details may be found in the [Mixpanel API Authentication](https://developer
...
    - `select_properties_by_default` (`true` or `false`): Mixpanel properties are not fixed and depend on the date being uploaded. During Discovery mode and catalog.json setup, all current/existing properties will be captured. Setting this config parameter to true ensures that new properties on events and engage records are captured. Otherwise new properties will be ignored.
    - `eu_residency_server` (`true` or `false`): Data Residency refers to the physical/geographical storage location of an organization's data or information. Setting this config parameter to true ensures that it uses eu_residency_server endpoint to capture the records. As a Mixpanel customer in the EU, you have the option to send your data to Mixpanel's EU data center, and have your data stored exclusively in the EU when creating a new project. [More info about eu_residency_server](https://help.mixpanel.com/hc/en-us/articles/360039135652-Data-Residency-in-EU).
+   - `request_timeout` (integer, `300`): Time for which request should wait to get response. Default request_timeout is 300 seconds.
```
Comment (dbshah1212) [inline README.md:141]:
```suggestion
   - `request_timeout` (integer, `300`): Max time for which request should wait to get a response. Default request_timeout is 300 seconds.
```

File: tap_mixpanel/__init__.py [hunk 4]
```diff
@@ -34,6 +35,11 @@ def main():
     parsed_args = singer.utils.parse_args(REQUIRED_CONFIG_KEYS)
 
     start_date = parsed_args.config['start_date']
+    # Set request timeout to config param `request_timeout` value.
+    # If value is 0,"0","" or not passed then it set default to 300 seconds.
```
Comment (dbshah1212) [inline tap_mixpanel/__init__.py:39]:
```suggestion
    # If value is 0, "0", "" or not passed then it sets default to 300 seconds.
```

File: tap_mixpanel/__init__.py [hunk 5]
```diff
@@ -34,6 +35,11 @@ def main():
...
+    config_request_timeout = parsed_args.config.get('request_timeout')
+    request_timeout = config_request_timeout and float(config_request_timeout) or REQUEST_TIMEOUT
```
Comment (dbshah1212) [inline tap_mixpanel/__init__.py:41]: Please change it to If-else to keep the consistency.

File: tap_mixpanel/streams.py [hunk 6]
```diff
@@ -285,6 +287,10 @@ def sync(self, state, catalog, config, start_date):
         days_interval = int(config.get("date_window_size", "30"))
         attribution_window = int(config.get("attribution_window", "5"))
 
+        # Set request timeout to config param `request_timeout` value.
+        # If value is 0,"0","" or not passed then it set default to 300 seconds.
```
Comment (dbshah1212) [inline tap_mixpanel/streams.py:291]:
```suggestion
        # If value is 0, "0", "" or not passed then it sets default to 300 seconds.
```

File: tap_mixpanel/streams.py [hunk 7]
```diff
@@ -285,6 +287,10 @@ def sync(self, state, catalog, config, start_date):
...
+        config_request_timeout = config.get('request_timeout')
+        request_timeout = config_request_timeout and float(config_request_timeout) or REQUEST_TIMEOUT
```
Comment (dbshah1212) [inline tap_mixpanel/streams.py:293]: Please change it to If-else to keep the consistency.

File: tests/configuration/fixtures.py [hunk 8]
```diff
@@ -4,6 +4,6 @@
 
 @pytest.fixture
 def mixpanel_client():
-    mixpanel_client = MixpanelClient('API_SECRET', api_domain="mixpanel.com")
+    mixpanel_client = MixpanelClient('API_SECRET', api_domain="mixpanel.com", request_timeout=1) # Pass extra request_timeout parameter
```
Comment (dbshah1212) [inline tests/configuration/fixtures.py:7]: Can you also add the reason of passing extra param in the comment?

File: tests/unittests/test_error_handling.py [hunk 9]
```diff
@@ -2,8 +2,11 @@
 from unittest import mock
```
Comment (dbshah1212) [inline tests/unittests/test_error_handling.py:2]: Please add comments for each test, like what it is testing.

File: tests/unittests/test_medium_client.py [hunk 10]
```diff
@@ -103,6 +103,19 @@ def test_request_backoff_on_timeout(mock_sleep, mixpanel_client):
...
+    Test that check_access method perform backoff on timeout error.
+    '''
+    with requests_mock.Mocker() as m:
+        m.request('GET', 'https://mixpanel.com/api/2.0/engage',
+                  exc=requests.exceptions.Timeout('Timeout on request'))
+
+        with raises(ReadTimeoutError) as ex:
+            result = mixpanel_client.check_access()
+        # Assert backoff retry count as expected
+        assert mock_sleep.call_count == 5 - 1
```
[the remark this answers is not in the document]
Comment (dbshah1212) [reply tests/unittests/test_medium_client.py:119]: Please write 4, No meaning of adding extra operation, Please elaborate the same in a comment with example.

File: tests/unittests/test_request_timeout_param_value.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    "start_date": "2020-02-01T00:00:00Z",
+    "user_agent": "tap-mixpanel <api_user_email@your_company.com>",
+    "eu_residency": False,
+    "end_date": "2020-03-02T00:00:00Z"
+}
+
+REQUEST_TIMEOUT = 300
+REQUEST_TIMEOUT_FLOAT = 300.0
+
+class MockCatalog():
```
Comment (dbshah1212) [inline tests/unittests/test_request_timeout_param_value.py:22]: Please add a comment for all mock classes.

File: tests/unittests/test_medium_client.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (prijendev) [reply tests/unittests/test_medium_client.py]: Updated to 4 and added a comment.

File: README.md (hunk 3 above)
Comment (prijendev) [reply README.md]: Updated

File: tap_mixpanel/__init__.py (hunk 4 above)
Comment (prijendev) [reply tap_mixpanel/__init__.py]: Updated

File: tap_mixpanel/streams.py (hunk 6 above)
Comment (prijendev) [reply tap_mixpanel/streams.py]: Updated

File: tap_mixpanel/streams.py (hunk 7 above)
Comment (prijendev) [reply tap_mixpanel/streams.py]: Updated

File: tests/configuration/fixtures.py (hunk 8 above)
Comment (prijendev) [reply tests/configuration/fixtures.py:8]: As support of request_timeout has been added, MixpanelClient accepts the request_timeout parameter which is mandatory.
Added comment in code also.

File: tests/unittests/test_error_handling.py (hunk 9 above)
Comment (prijendev) [reply tests/unittests/test_error_handling.py:2]: Added

File: tests/unittests/test_request_timeout_param_value.py (hunk 11 above)
Comment (prijendev) [reply tests/unittests/test_request_timeout_param_value.py:22]: Added

**Review by karanpanchal-crest — CHANGES_REQUESTED** [review]
There is a scope of optimization in the code @prijendev .
Please incorporate the changes as discussed on the call

**Review by savan-chovatiya — APPROVED** [review]

**Review by karanpanchal-crest — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**